### PR TITLE
[CORE][VL] Fix memory leak of fallback storage

### PR DESCRIPTION
**Pull request by Yohahaha**
## What changes were proposed in this pull request?

Before this patch, the resources which not tracked by Spark executor task will be managed in FALLBACK_REGISTRY(java side) and fallbackStorage(native side), however, these will be released when process exit, and may bring memory problem when running lots of query in one Spark application.

E.g. DPP subquery execution will submitted to another thread pool in Spark driver, Gluten need do c2r conversion in the same time. We add explicit release after conversion.
Some computation happens on another thread, such as PythonUDF, since Velox backend has not support PythonUDF, it will insert a c2r, this c2r will do conversion in PythonRunner writer thread. We add a check in c2r to find this case and release early.

### CODE REVIEW

**Comment by PHILO-HE** [comment]
@zhztheplayer, could you please take a review? Thanks.

**Comment by zhztheplayer** [comment]
@Yohahaha 
Is this ready to review? Or let's see if the python test suites can pass.

**Comment by Yohahaha** [comment]
> @Yohahaha Is this ready to review? Or let's see if the python test suites can pass.

Yes, you can start review, python suites has been verified in my machine.

**Comment by Yohahaha** [comment]
> > @Yohahaha Is this ready to review? Or let's see if the python test suites can pass.
> 
> Yes, you can start review, python suites has been verified in my machine.

Not ready... Modification of TaskResources is not enough.

**Comment by Yohahaha** [comment]
@zhztheplayer this PR is ready for review, I have run whole tests in local.

File: cpp/core/utils/TaskContext.cc [hunk 1]
```diff
@@ -40,6 +40,8 @@ class TaskContextStorage {
 };
 
 thread_local std::unique_ptr<TaskContextStorage> taskContextStorage = nullptr;
+
+// TODO: Remove this and let related unmanaged resources be managed.
```
Comment (jackylee-ch) [inline cpp/core/utils/TaskContext.cc:44]: So we still have some memory unmanaged, right?

**Comment by zhztheplayer** [comment]
> > @Yohahaha Is this ready to review? Or let's see if the python test suites can pass.
> 
> Yes, you can start review, python suites has been verified in my machine.

Thanks, I'll take a look ASAP.

File: cpp/core/utils/TaskContext.cc (hunk 1 above)
Comment (Yohahaha) [reply cpp/core/utils/TaskContext.cc:44]: Please ignore this obsolete comments and will clean later, I add false check when using fallback storage, ideally all native resources in under managed.

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarToRowExec.scala [hunk 2]
```diff
@@ -134,6 +135,8 @@ class ColumnarToRowRDD(
             if (!hasNext && !closed) {
               jniWrapper.nativeClose(c2rId)
               closed = true
+
+              TaskResources.releaseCurrentResources()
```
Comment (zhztheplayer) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarToRowExec.scala:139]: Would you like to explain a bit about this? Thanks!

Comment (zhztheplayer) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarToRowExec.scala:139]: If there is a case: (G for Gluten, V for Vanilla Spark)

```
G_ShuffleRead -> C2R_1 -> V_Plan_1 -> R2C_1 -> G_Plan_1 -> C2R_2 -> V_ShuffleWrite
```

Once C2R_1 reached EoS, C2R_2 might still consume data. Is the case covered by the code?

Comment (Yohahaha) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarToRowExec.scala:139]: Sure, the key changes is `RESOURCE_REGISTRIES` which contains mapping of 
(TaskContext -> (ThreadID -> TaskResourceRegistry)) now, to handle different thread computation under same Spark TaskContext scope, let it could be released.

If all computations are offloaded to native, there should no multi computation under same Spark task scope. If we fallback some special operators, c2r will be inserted, and this parts of computation could happens in another thread, such as PythonRunner, so we need add extra mapping to release resources created in another thread.

Comment (Yohahaha) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarToRowExec.scala:139]: > Once C2R_1 reached EoS, C2R_2 might still consume data. Is the case covered by the code?

Yes, I add a UnsafeRow#copy at the end of ColumnarToRowRDD.

File: gluten-core/src/main/scala/org/apache/spark/util/TaskResources.scala [hunk 3]
```diff
@@ -60,20 +51,97 @@ object TaskResources extends TaskListener with Logging {
...
+  }
+
+  def releaseCurrentResources(): Unit = {
+    RESOURCE_REGISTRIES.synchronized {
+      val currentTId = Thread.currentThread().getId
+      val registryMap = RESOURCE_REGISTRIES.get(getLocalTaskContext())
+      if (registryMap.containsKey(currentTId) && registryMap.get(currentTId).needExplicitRelease) {
+        val registry = registryMap.remove(currentTId)
+        logDebug(s"Release all resources in ${registry.name}")
+        registry.releaseAll()
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/org/apache/spark/util/TaskResources.scala:143]: So by the condition `needExplicitRelease`, `releaseCurrentResources` only releases the resources from off-task thread, does it?

Comment (Yohahaha) [reply gluten-core/src/main/scala/org/apache/spark/util/TaskResources.scala:143]: Yes.

File: cpp/core/utils/TaskContext.cc [hunk 4]
```diff
@@ -55,13 +52,7 @@ void bindToTask(std::shared_ptr<void> object) {
...
     return;
   }
-  // The fallback storage is used. Spark sometimes creates sub-threads from a task thread. For example,
-  //   PythonRunner.scala:183 @ Spark3.2.2
-  //   GlutenSubqueryBroadcastExec.scala:73 @6bce2c33
-  std::lock_guard<std::mutex> guard(fallbackStorageMutex);
-  std::cout << "Binding a shared object to fallback storage. This should only happen on sub-thread of a Spark task. "
-            << std::endl;
-  fallbackStorage->bind(object);
+  GLUTEN_CHECK(false, "[BUG] Found untracked computation, please submit issue with query plan.");
```
Comment (zhztheplayer) [inline cpp/core/utils/TaskContext.cc:55]: This fix is not related to the Java side [TaskContext, [TID, Resource]] design?

Comment (Yohahaha) [reply cpp/core/utils/TaskContext.cc:55]: Just put check here to find corner cases we ignored.

**Comment by zhztheplayer** [comment]
@Yohahaha Sorry for the slow response but let's find some time to have a dedicated sync for this change. Thanks.

**Comment by Yohahaha** [comment]
> @Yohahaha Sorry for the slow response but let's find some time to have a dedicated sync for this change. Thanks.

Sure, we could have a phone call by wechat.

**Comment by Yohahaha** [comment]
PR #2843 is better way to solve this issue.
